### PR TITLE
Identity REST authenticator with MSF4J interceptor

### PR DESCRIPTION
**Pull request by ruwanta**
Redmine issue: 
https://redmine.wso2.com/issues/5856

### CODE REVIEW

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/test/java/org/wso2/carbon/identity/auth/msf4j/internal/SimpleTest.java [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.ws.rs.HttpMethod;
+
+public class SimpleTest {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/test/java/org/wso2/carbon/identity/auth/msf4j/internal/SimpleTest.java]: Add a class comment.

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/test/java/org/wso2/carbon/identity/auth/msf4j/internal/SimpleTest.java [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        //        tester.call();
+        tester.callSuccess();
+        tester.callFail();
+
+        runner.stop();
+    }
+
+    protected void call() throws IOException {
+        HttpURLConnection urlConn = request("/hello/me", HttpMethod.GET, false);
+        System.out.println(urlConn.getResponseCode());
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/test/java/org/wso2/carbon/identity/auth/msf4j/internal/SimpleTest.java]: Remove `sout`.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationResult.java [hunk 3]
```diff
@@ -18,15 +18,20 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 /**
  * AuthenticationResult hold the status.
  */
-public class AuthenticationResult implements Serializable {
+public class AuthenticationResult{
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationResult.java]: Formatting issue.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationResult.java [hunk 4]
```diff
@@ -48,4 +53,23 @@ public AuthenticationStatus getAuthenticationStatus() {
...
+    public int getStatusCode() {
+        return statusCode;
+    }
+
+    public void setStatusCode(int statusCode) {
+        this.statusCode = statusCode;
+    }
+
+    public List<Header> getResponseHeaders() {
+        return responseHeaders == null? Collections.emptyList() : Collections.unmodifiableList(responseHeaders);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationResult.java]: Formatting issue.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        }
+        return false;
+    }
+
+    /**
+     * Returns the domain name from the user string.
+     * @param userName
+     * @return
+     */
+    protected ImmutablePair<String, String> decodeTenantDomainAndUserName(String userName) {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java]: The method is misleading as it is not related to tenant domains.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        if(userName != null && userName.contains("/") ) {
+            int lastAt = userName.lastIndexOf('/');
+            domainName = userName.substring(0,lastAt);
+            userName = userName.substring(lastAt + 1);
+        }
+        if(domainName != null) {
+            domainName = domainName.toLowerCase(Locale.getDefault());
+        }
+        if(domainName == null) {
+            domainName = "PRIMARY";
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java]: We cannot guarantee that the primary domain name will be "PRIMARY". If needed we can either use `identityStore.getPrimaryDomain()`.

If the domain is not available, is it reasonable to assume the domain is primary ? If not, we can just pass `null` as the domain, so the authentication will happen across all domains.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java [hunk 7]
```diff
@@ -18,152 +18,136 @@
...
+                                "Could not get the identity store to autnenticate the user with "
+                                        + "BASIC authentication. Username: "
+                                        + userName;
+                        if (log.isDebugEnabled()) {
+                            log.debug(errorMessage);
+                        }
+                        throw new AuthenticationFailException(errorMessage);
+                    } else {
+
+                        Claim userNameClaim = new Claim("http://wso2.org/claims", "http://wso2.org/claims/username",
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java]: Use the constants, `IdentityMgtConstants.CLAIM_ROOT_DIALECT` and `IdentityMgtConstants.USERNAME_CLAIM`

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java [hunk 8]
```diff
@@ -18,152 +18,136 @@
...
+                                userName);
+                        PasswordCallback passwordCallback = new PasswordCallback("password", false);
+                        passwordCallback.setPassword(password);
+                        org.wso2.carbon.identity.mgt.AuthenticationContext resultAuthenticationContext = identityStore
+                                .authenticate(userNameClaim, new Callback[] { passwordCallback }, domainName);
+                        if (resultAuthenticationContext.getUser() != null) {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java]: Can check `resultAuthenticationContext.isAuthenticated()`.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java [hunk 9]
```diff
@@ -18,152 +18,136 @@
...
                     }
-                } catch ( org.wso2.carbon.user.api.UserStoreException e ) {
-                    String errorMessage = "Error occurred while trying to authenticate, " + e.getMessage();
-                    log.error(errorMessage);
-                    throw new AuthenticationFailException(errorMessage);
-                } finally {
-                    PrivilegedCarbonContext.endTenantFlow();
+                } catch (AuthenticationFailure authenticationFailure) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Authentication failed, with the user: "+userName, authenticationFailure);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java]: Formatting issue.

**Comment by omindu** [comment]
Better to run a build with checkstyle enabled and mvn javadoc:javadoc.

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/pom.xml [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            org.wso2.carbon.identity.auth.service.*;version="${org.wso2.carbon.identity.auth.service.version.range}",
+            org.wso2.msf4j.*;version="${msf4j.package.import.version.range}",
+            org.slf4j.*;version="${slf4j.logging.package.import.version.range}",
+            javax.ws.rs.core,
+            org.wso2.carbon.messaging;
+        </import.package>
+        <export.package>
+            org.wso2.carbon.identity.auth.msf4j.*;version="${project.version}"
+        </export.package>
+        <dynamic.import.package>*</dynamic.import.package>
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/pom.xml]: Removed dynamic import

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.auth.msf4j.internal;
+
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.*;
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java]: Done

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java (hunk 9 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java]: Fixed

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+import org.slf4j.LoggerFactory;
+import org.wso2.carbon.identity.auth.service.*;
+import org.wso2.carbon.identity.auth.service.exception.AuthClientException;
+import org.wso2.carbon.messaging.Header;
+import org.wso2.msf4j.Interceptor;
+import org.wso2.msf4j.Request;
+import org.wso2.msf4j.Response;
+import org.wso2.msf4j.ServiceMethodInfo;
+
+import java.net.HttpCookie;
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java]: Handled Cookie parsing with custom logic, without any additional libs. Tests are added.

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        AuthenticationContext authenticationContext = new AuthenticationContext(authenticationRequest);
+        AuthenticationResult authenticationResult = authenticationManager.authenticate(authenticationContext);
+        if (authenticationResult.getAuthenticationStatus() == AuthenticationStatus.SUCCESS) {
+            /*The authzUser is a workaround since MSF4J does not have proper SecurityContext handling
+            * This has to be removed and associate the proper Principal Object to make it similar to
+            * JAX-RS way to get the user.*/
+            request.setProperty("authzUser", authenticationContext.getUser().getUniqueUserId());
+            return true;
+        }
+        if(authenticationResult.getStatusCode() >0) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java]: Fixed

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        if (log.isDebugEnabled()) {
+            log.debug("PostCall made for the service Method: %s, Path: %s", serviceMethodInfo.getMethod(),
+                    request.getUri());
+        }
+    }
+
+    private AuthenticationRequest.AuthenticationRequestBuilder createRequestBuilder(Request request, Response response)
+            throws AuthClientException {
+
+        AuthenticationRequest.AuthenticationRequestBuilder authenticationRequestBuilder = new AuthenticationRequest.AuthenticationRequestBuilder();
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java]: Fixed

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+        if (request.getProperties() != null) {
+
+            Set<String> propertyNames = request.getProperties().keySet();
+
+            for (String propertyName : propertyNames) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java]: Done

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+            headers.forEach(header -> authenticationRequestBuilder.addHeader(header.getName(), header.getValue()));
+
+        }
+
+        String cookieHeader = request.getHeader(HttpHeaders.COOKIE);
+        List<HttpCookie> cookies = Collections.emptyList();
+        if (cookieHeader != null) {
+            cookies = HttpCookie.parse(cookieHeader);
+        }
+        for (HttpCookie cookie : cookies) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/main/java/org/wso2/carbon/identity/auth/msf4j/internal/AuthenticationInterceptor.java]: Changed with custom cookie parsing.

File: components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/test/java/org/wso2/carbon/identity/auth/msf4j/internal/SimpleTest.java (hunk 1 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.msf4j.interceptor/src/test/java/org/wso2/carbon/identity/auth/msf4j/internal/SimpleTest.java]: Removed this unnecessary class.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationResult.java (hunk 3 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationResult.java]: Fixed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java (hunk 8 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java]: Changed according to the comment.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java (hunk 7 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/BasicAuthenticationHandler.java]: Changed as per the comment.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java (hunk 6 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java]: Changed to pass null when no domain provided.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java (hunk 5 above)
Comment (ruwanta) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java]: Changed the method signature to better describe the purpose.

**Comment by JKAUSHALYA** [comment]
Closing until conflicts are resolved.

**Pull request closed without merging**